src/overlay/OverlayManagerImpl.cpp

```diff
@@ -1182,54 +1199,109 @@ OverlayManagerImpl::checkScheduledAndCache(
+    {
+        forgetFloodedMsg(index);
+        CLOG_DEBUG(Overlay,
+                   "Peer::recvTransaction Discarded transaction {} from {}",
+                   hexAbbrev(transaction->getFullHash()), peer->toString());
+    }
+    else
+    {
+        bool dup =
+            addCode == TransactionQueue::AddResultCode::ADD_STATUS_DUPLICATE;
+        if (!dup)
+        {
+            pulledRelevantTx = true;
+        }
+        CLOG_DEBUG(Overlay,
+                   "Peer::recvTransaction Received {} transaction {} from {}",
+                   (dup ? "duplicate" : "unique"),
+                   hexAbbrev(transaction->getFullHash()), peer->toString());
+    }
+
+    auto const& om = getOverlayMetrics();
+    auto& meter =
+        pulledRelevantTx ? om.mPulledRelevantTxs : om.mPulledIrrelevantTxs;
+    meter.Mark();
+}
+
 void
 OverlayManagerImpl::recvTransaction(StellarMessage const& msg,
-                                    Peer::pointer peer, Hash const& index)
+                                    Peer::pointer peer,
+                                    std::shared_ptr<Hash const> index)
 {
     ZoneScoped;
+    releaseAssert(index != nullptr);
     auto transaction = TransactionFrameBase::makeTransactionFromWire(
         mApp.getNetworkID(), msg.transaction());
     if (transaction)
     {
         // record that this peer sent us this transaction
         // add it to the floodmap so that this peer gets credit for it
-        recvFloodedMsgID(msg, peer, index);
+        recvFloodedMsgID(msg, peer, *index);
 
         mTxDemandsManager.recordTxPullLatency(transaction->getFullHash(), peer);
 
-        // add it to our current set
-        // and make sure it is valid
-        auto addResult = mApp.getHerder().recvTransaction(transaction, false);
-        bool pulledRelevantTx = false;
-        if (!(addResult.code ==
-                  TransactionQueue::AddResultCode::ADD_STATUS_PENDING ||
-              addResult.code ==
-                  TransactionQueue::AddResultCode::ADD_STATUS_DUPLICATE))
+        if (mApp.getConfig().BACKGROUND_TX_VALIDATION)
         {
-            forgetFloodedMsg(index);
-            CLOG_DEBUG(Overlay,
-                       "Peer::recvTransaction Discarded transaction {} from {}",
-                       hexAbbrev(transaction->getFullHash()), peer->toString());
+            auto closeTime = mApp.getLedgerManager()
+                                 .getLastClosedLedgerHeader()
+                                 .header.scpValue.closeTime;
+            auto offset =
+                getUpperBoundCloseTimeOffset(mApp.getAppConnector(), closeTime);
+            mApp.postOnTxValidationThread(
+                [this, transaction, offset, peer, index]() {
+                    ZoneScopedN("background checkValid");
+                    LOCK_GUARD(mSnapshotMutex, guard);
+                    releaseAssert(mLedgerSnapshot);
+
+                    auto txResult =
```

Comment:
I think we're missing some tests: for transaction validation purposes, ledger state needs to be slightly modified, similar to how we do it in `TransactionQueue::canAdd`
```
    if (protocolVersionStartsFrom(ledgerVersion, ProtocolVersion::V_19))
    {
        // This is done so minSeqLedgerGap is validated against the next
        // ledgerSeq, which is what will be used at apply time
        ls.getLedgerHeader().currentToModify().ledgerSeq =
            mApp.getLedgerManager().getLastClosedLedgerNum() + 1;
    }
```